### PR TITLE
enhance: disk cache fault tolerance

### PR DESCRIPTION
**Pull request by jiefenghuang**

### CODE REVIEW

File: pkg/chunk/disk_cache_state.go [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+			cache.state.stop()
+			cache.state = newDCState(dcDown, cache)
+		}
+	}
+	logger.Warnf("disk cache state change from %s to %s", diskStateNames[state], diskStateNames[cache.state.state()])
+}
+
+func getEnvs() {
+	if os.Getenv("JFS_MAX_IO_DURATION") != "" {
+		dur, _ := strconv.Atoi(os.Getenv("JFS_MAX_IO_DURATION"))
```
Comment (SandyXSD) [inline pkg/chunk/disk_cache_state.go:257]: Don't alter the argument if the env is invalid?

File: pkg/chunk/disk_cache_state.go [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+		maxIODur = time.Duration(dur) * time.Second
+		logger.Infof("set maxIODur to %d", maxIODur)
+	}
+	if os.Getenv("JFS_MAX_IO_ERR_PERCENTAGE") != "" {
+		percentage, _ := strconv.ParseFloat(os.Getenv("JFS_MAX_IO_ERR_PERCENTAGE"), 64)
+		maxIOErrPercentageToNormal = 1 - percentage
```
Comment (SandyXSD) [inline pkg/chunk/disk_cache_state.go:263]: remove this line?

File: pkg/chunk/disk_cache_state.go [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+		logger.Infof("set maxIOErrPercentageToNormal to %f", maxIOErrPercentageToNormal)
+	}
+	if os.Getenv("JFS_MAX_DURATION_TO_DOWN") != "" {
+		dur, _ := strconv.Atoi(os.Getenv("JFS_MAX_DURATION_TO_DOWN"))
+		maxDurToDown = time.Duration(dur) * time.Minute
```
Comment (SandyXSD) [inline pkg/chunk/disk_cache_state.go:268]: better support duration string, like `1m`

File: pkg/chunk/disk_cache_state.go [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+		ticker := time.NewTicker(tickDurForUnstable)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-dc.stopCh:
+				return
+			case <-ticker.C:
+				errCnt, ioCnt := atomic.LoadUint32(&dc.ioErrCnt), atomic.LoadUint32(&dc.ioCnt)
+				if ioCnt >= minIOSuccToNormal && float64(errCnt)/float64(ioCnt) <= maxIOErrPercentageToNormal {
```
Comment (SandyXSD) [inline pkg/chunk/disk_cache_state.go:173]: `minIOSuccToNormal` too big?

File: pkg/chunk/disk_cache_state.go [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+					dc.cache.event(eventToNormal)
+				} else if time.Since(dc.startTime) >= maxDurToDown {
+					dc.cache.event(eventToDown)
+				} else {
+					atomic.StoreUint32(&dc.ioErrCnt, 0)
```
Comment (SandyXSD) [inline pkg/chunk/disk_cache_state.go:178]: timeout ops will be recorded for only once.

File: pkg/chunk/disk_cache_state_test.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
+package chunk
```
Comment (davies) [inline pkg/chunk/disk_cache_state_test.go:1]: license header

File: pkg/chunk/disk_cache_state.go [hunk 7]
```diff
@@ -0,0 +1,297 @@
+package chunk
```
Comment (davies) [inline pkg/chunk/disk_cache_state.go:1]: license header

File: pkg/chunk/disk_cache_state.go [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+		switch eventType {
+		case eventToNormal:
+			cache.state.stop()
+			cache.state = newDCState(dcNormal, cache)
+		case eventToDown:
+			cache.state.stop()
+			cache.state = newDCState(dcDown, cache)
+		}
+	}
+	logger.Warnf("disk cache state change from %s to %s", diskStateNames[state], diskStateNames[cache.state.state()])
```
Comment (davies) [inline pkg/chunk/disk_cache_state.go:257]: add path into logging

Comment (davies) [reply pkg/chunk/disk_cache_state.go:257]: info

File: pkg/chunk/disk_cache_state.go [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+		select {
+		case <-dc.stopCh:
+			return
+		default:
+			cnt++
+			start := time.Now()
+			dc.doProbe(probeCacheKey(cnt, len(probeData)), page)
+			diff := probeDur - time.Since(start)
+			if diff > 0 {
+				time.Sleep(diff)
```
Comment (davies) [inline pkg/chunk/disk_cache_state.go:203]: 可以再久一点，减少IO 消耗

Comment (jiefenghuang) [reply pkg/chunk/disk_cache_state.go:203]: change: probeDur from 100ms to 500ms (<= 2 qps), minIOSuccToNormal from 3*60 to 60(>= 1 qps)

File: pkg/chunk/disk_cache_state.go (hunk 7 above)
Comment (jiefenghuang) [reply pkg/chunk/disk_cache_state.go:1]: done

File: pkg/chunk/disk_cache_state.go (hunk 8 above)
Comment (jiefenghuang) [reply pkg/chunk/disk_cache_state.go:257]: done

File: pkg/chunk/disk_cache_state_test.go (hunk 6 above)
Comment (jiefenghuang) [reply pkg/chunk/disk_cache_state_test.go:1]: done

File: pkg/chunk/disk_cache_state.go (hunk 8 above)
Comment (SandyXSD) [reply pkg/chunk/disk_cache_state.go]: @jiefenghuang logger level?

Comment (jiefenghuang) [reply pkg/chunk/disk_cache_state.go]: we had error msg when removeStore in cacheManager, so i think warning is enough

Comment (jiefenghuang) [reply pkg/chunk/disk_cache_state.go]: @SandyXSD changed to info

File: pkg/chunk/disk_cache_state.go (hunk 9 above)
Comment (davies) [reply pkg/chunk/disk_cache_state.go:219]: 没问题了，之前看错了，以为  diff 是测试消耗的时间

**Comment by jiefenghuang** [comment]
close # 4798

**Comment by jiefenghuang** [comment]
close #4798

**Pull request merged**